Remove proxymap entry after closing connection

tgraf opened the pull request.

tgraf committed:
proxymap: Add generic interface to delete, lookup and clean entries

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kafka: Remove proxymap entries on closing of receive connection

When the receive side of the proxied connection is closed, after the SO_LINGER
timeout has passed, remove the proxymap entry as we no longer need access to
the metadata. This relaxes the garbage collector and pressure on the map
capacity.

Fixes: #3172

Signed-off-by: Thomas Graf <thomas@cilium.io>

rlenglet commented on pkg/proxy/kafka.go:
```diff
@@ -356,6 +356,14 @@ func (k *kafkaRedirect) handleRequestConnection(pair *connectionPair) {
+	// The proxymap contains an entry with metadata for the receive side of the
+	// connection, remove it after the connection has been closed.
+	if pair.Rx != nil {
+		if err := k.redirect.removeProxyMapEntryOnClose(pair.Rx.conn); err != nil {
```
It can take up to 1 minute for the rx connection to get actually closed, because that’s the delay we set for SO_LINGER. Therefore, we need to wait for at least 1 minute before removing the proxy map entry. Otherwise, the connection may still hang for the client.

tgraf replied:
> If 1 minute is too long, we can reduce the SO_LINGER delay. But we should always wait for at least that delay before removing the map entry.

Good point. I'm not sure lingering is actually better though. 1 minute is definitely too long just to send out some existing data. I agree that we should wait though.

tgraf committed:
proxy: Reduce SO_LINGER timeout to 10 seconds

The existing value of 1 minute can put a lot of stress on the proxymap table in
an environment with many concurrent, short lived connections.

Signed-off-by: Thomas Graf <thomas@cilium.io>

rlenglet commented on pkg/proxy/kafka.go:
```diff
@@ -356,6 +356,19 @@ func (k *kafkaRedirect) handleRequestConnection(pair *connectionPair) {
+		// We are running in our own go routine here so we can just
+		// block this go routine until after the connection is
+		// guaranteed to have been closed
+		time.Sleep(proxyConnectionCloseTimeout)
```
I would add a small buffer, to let an eventual RST packet get through. In the worst case, the IP stack will decide to send an RST just around that time. We need to give it a chance to get through. How about adding like 1 second?
```
time.Sleep(proxyConnectionCloseTimeout + 1 * time.Second)
```

tgraf replied:
Done, can't hurt. It's probably unlikely though that we can transmit an RST if we can't flush out the data and close the connection within the timeout. It's likely a connectivity problem anyway.

rlenglet reviewed (approved):
LGTM. I would just add a small buffer to the sleep.
